src/solvers/mp/MasterProblem.h

```diff
@@ -0,0 +1,386 @@
+
+      return true;
+    };
+
+    virtual std::string toString(int nbDays = -1, std::vector<int> shiftIDToShiftTypeID={}) const {
+      return "";
+    }
+
+    virtual int getShift(int day) const = 0;
+
+    // need to be able to write the pattern as a vector and to create a new one from it
+    virtual std::vector<double> getCompactPattern() const {
+      std::vector<double> compact;
+      compact.push_back(nurseId_);
+      compact.push_back(firstDay_);
+      compact.push_back(length_);
+      return compact;
+    }
+
+    int nurseId_;
+    int firstDay_;
+    int length_;
+
+};
+
+//-----------------------------------------------------------------------------
+//
+//  C l a s s   M a s t e r P r o b l e m
+//
+// Build and solve the master problem of the column generation scheme
+//
+//-----------------------------------------------------------------------------
+
+enum MySolverType { S_SCIP, S_CLP, S_Gurobi, S_Cplex, S_CBC };
+static std::map<std::string,MySolverType> MySolverTypesByName =
+{{"CLP",S_CLP},{"Gurobi",S_Gurobi},{"Cplex",S_Cplex},{"CBC",S_CBC},{"SCIP",S_SCIP}};
+
+class MasterProblem : public Solver, public PrintSolution{
+  public:
+    // Specific constructor and destructor
+    MasterProblem(Scenario* pScenario, Demand* pDemand,
+      Preferences* pPreferences, std::vector<State>* pInitState, MySolverType solver);
+    ~MasterProblem();
+
+    //solve the rostering problem
+    double solve(std::vector<Roster> solution = {});
+
+    //solve the rostering problem or just the relaxation(root node)
+    double solve(std::vector<Roster> solution, bool rebuild);
+
+    // Solve with parameters
+    double solve(const SolverParam& parameters, std::vector<Roster> solution = {});
+
+    //Resolve the problem with another demand and keep the same preferences
+    //
+    double resolve(Demand* pDemand, const SolverParam& parameters, std::vector<Roster> solution = {});
+
+    // needs to be specialized: add a colum  to the master from a solution of the subproblem
+    virtual MyVar* addColumn(int nurseId, const RCSolution& solution) = 0;
+
+    // retrieve the object represented ny the  vector pattern
+    virtual PPattern getPattern(const std::vector<double>& pattern) const = 0;
+
+    // throw an error if pattern is already present as an active column
+    void checkIfPatternAlreadyPresent(const std::vector<double>& pattern) const;
+
+    virtual const std::vector<MyVar*>& getRestVarsPerDay(LiveNurse* pNurse, int day) const = 0;
+
+    //get the pointer to the model
+    Modeler* getModel(){
+      return pModel_;
+    }
+
+    //override PrintSolution virtual method
+    void save(std::vector<int>& weekIndices, std::string outdir);
+    void printCurrentSol();
+
+    // build the, possibly fractional, roster corresponding to the solution
+    // currently stored in the model
+    vector3D<double> getFractionalRoster() ;
+
+    // build a DualCosts structure
+    DualCosts buildDualCosts(LiveNurse* pNurse) const;
```

Comment:
Dual costs are dependent on the type of master problem. This should be in subclass, shouldn'it?

Reply:
The idea is the structure of the dual costs is rigid, but the functions to populate it are virtual.

Reply:
Here is the code of buildDualCosts: 
`DualCosts MasterProblem::buildDualCosts(LiveNurse*  pNurse) const {
  return DualCosts(getShiftsDualValues(pNurse), getStartWorkDualValues(pNurse),
      getEndWorkDualValues(pNurse), getWorkedWeekendDualValue(pNurse));
}`
My issue is that most of the populate functions do not even make sense in a decomposition by rotation. 
Actually, I am even wondering whether we should not try and make the DualCosts class more general. E.g., instead of referring to : 
` // Indexed by : (day, shift) !! 0 = shift 1 !!
  vector2D<double> workedShiftsCosts_;

  // Indexed by : day
  std::vector<double> startWorkCosts_;

  // Indexed by : day
  std::vector<double> endWorkCosts_;

  // Reduced cost of the weekends
  double workedWeekendCost_;`
refer only to dual costs on (day,shift)s and on days: `dualDays_, dualShifts_`. For the rotation model, build `dualDays_ startWorkCosts_+endWorkCosts_ + (workedWeekendCost_ if day on week-end)`

Could make an issue if you agree.